source/RoslynAnalyzers/ApiControllerAnalyzer.cs

```diff
@@ -0,0 +1,160 @@
+            if (symCmp.Equals(typeSymbol, context.Compilation.GetTypeByMetadataName("Octopus.Server.Web.Controllers.Telemetry.AmplitudeOkResult"))) return true;
+         
+            // FileResult plus derived classes FileContentResult and FileStreamResult
+            var fileResult = context.Compilation.GetTypeByMetadataName("Microsoft.AspNetCore.Mvc.FileResult");
+            if (symCmp.Equals(typeSymbol, fileResult) || typeSymbol.InheritsFrom(fileResult)) return true;
+            
+            // only CreatedResult<T> is allowed, the non-generic one isn't
+            if (symCmp.Equals(typeSymbol.OriginalDefinition, context.Compilation.GetTypeByMetadataName("Octopus.Server.Extensibility.Web.Extensions.CreatedResult`1"))) return true; 
+            
+            context.ReportDiagnostic(Diagnostic.Create(MustNotReturnActionResults, location: methodDec.ReturnType.GetLocation()));
+            return false;
+        }
+        
+        return true;
+    }
+
+    /// <summary>
+    /// Returns a guess at whether a method on a controller is an MVC action method.
+    /// </summary>
+    static bool IsActionMethod(SyntaxNodeAnalysisContext context, MethodDeclarationSyntax methodDec)
+    {
+        var isPublic = false;
+        foreach (var s in methodDec.Modifiers)
+        {
+            if (s.IsKind(SyntaxKind.PublicKeyword)) isPublic = true;
+
+            if (s.IsKind(SyntaxKind.StaticKeyword)) return false;
+            if (s.IsKind(SyntaxKind.AbstractKeyword)) return false;
+            if (s.IsKind(SyntaxKind.ConstructorDeclaration)) return false;
+        }
+
+        if (!isPublic) return false;
+
+        var isObviouslyActionMethod = false;
+        foreach (var attr in methodDec.AttributeLists.SelectMany(al => al.Attributes))
+        {
+            if(attr.Name is not IdentifierNameSyntax { Identifier.Text: var identifierText }) continue;
+
+            switch (identifierText)
+            {
+                case "Route":
+                case "HttpGet":
+                case "HttpPost":
+                case "HttpPut":
+                case "HttpDelete":
+                    isObviouslyActionMethod = true;
+                    break;
+            }
+        }
+        
+        // TODO someone could have made a custom subclass of HttpGet, we'd need to walk the inheritance tree here to catch it.
+        // TODO that work goes here later.
+
+        return isObviouslyActionMethod;
+    }
+
+    static bool IsApiController(SyntaxNodeAnalysisContext context, ClassDeclarationSyntax classDec)
+    {
+        var isPublic = false;
+        foreach (var s in classDec.Modifiers)
+        {
+            if (s.IsKind(SyntaxKind.PublicKeyword)) isPublic = true;
+
+            if (s.IsKind(SyntaxKind.StaticKeyword)) return false;
+            if (s.IsKind(SyntaxKind.AbstractKeyword)) return false;
+        }
+        if(!isPublic) return false;
```

Comment:
Might be morning brain but I can't see why we evaluate isPublic outside of the loop, but the others return inside the loop.

Reply:
because public might appear at any position in the array, and we can't be sure we've found it until we iterate the entire thing

Reply:
Ah of course, its because we are looking for absence :) morning brain indeed.